### DISCUSSION TITLE
🗺️ Remix SPA Target

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#7638] [Official RFCs]
# Remix SPA

While a React Router SPA and Remix feel very similar, there's still a decent gap between them from a developer ergonomics perspective. This proposal closes that gap allowing Remix to function as a Single Page App.

Additionally, React Router apps don't have a fluid migration path to Remix today. Every loader and action needs to suddenly start working on the server, file system routes need to be figured out from whatever the app was doing before, etc. Along with the #7632 proposal, and the #7634 proposal, this proposal creates a fluid migration path from React Router to Remix. Developers can add a JavaScript backend to their Remix app without changing application code outside of the client entry point.

## Proposal

In the vite config, a flag changes the mode to "spa" from the default "ssr".

```tsx
import { defineConfig } from "vite";
import { unstable_remix } from "@remix-run/vite";

export default defineConfig({
  plugins: [
    unstable_remix({
      // Change the target to "spa"
      target: "spa", // "ssr" | "spa"
    }),
  ],
});
```

- The remix plugin will now only generate client bundles, and normal vite behavior will create the index.html entry point to the app.

- It will use the same file conventions for `entry.client.tsx`, `root.tsx`, and file system routes.

- The route module API remains the same, but the `loader` and `action` exports will be ignored, and only the `clientLoader` and `clientAction` exports will be used.

  ```tsx
  export async function clientLoader({ request, params }) {
    let data = await fetch("/api/data");
    return { data };
  }

  export function clientAction({ request, params }) {
    await saveStuff(request);
    return { ok: true };
  }
  ```

- Imports still come from `@remix-run/react` not `react-router-dom`

  ```tsx
  import { NavLink } from "@remix-run/react";
  ```

### Entry points

The HTML entry point for vite is defined at `app/entry.html` with a script pointing to `entry.client.tsx`:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Remix SPA</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/entry.client.tsx"></script>
  </body>
</html>
```

The app `entry.client.tsx` will need to look a little different than ssr Remix by rendering into a root element instead of hydrating a full document:

```tsx
import * as React from "react";
import * as ReactDOM from "react-dom/client";
import { RemixBrowser } from "@remix-run/react";

ReactDOM.createRoot(document.getElementById("root")).render(
  <RemixBrowser />
);
```

## Migrating from SPA to SSR targets

To switch to a full stack Remix app with the Remix app server, you simply change the target, but everything else remains the same:

```diff
export default defineConfig({
  plugins: [
    unstable_remix({
-      target: "spa",
+      target: "ssr",
    }),
  ],
});
```

To switch to a custom server, apps will also need to add a server entry file and adapter to the vite config. That setup is discussed in #7632 proposal.

**Comment by ngbrown** [comment]
Instead of ignoring the `loader` and `action` exports when in SPA mode, should errors be thrown? Or at least an eslint warning triggered?

**Reply by ryanflorence** [reply]
yeah, ignored at runtime, but perhaps at build time we can warn if there's a loader/action with a spa target.

**Reply by rossipedia** [reply]
I would assume that this would slot in nicely with the `clientLoader`/`clientAction` proposal? ie: those would just roughly translate to RR `loader`/`action`?

**Reply by rossipedia** [reply]
Just caught the callout at the end of the client data proposal, question answered :+1:

**Comment by estevaolucas** [comment]
That's great! SPA target will probably unlock us to use Remix for micro-frontends powered by [Single-Spa](https://github.com/single-spa/single-spa/) + Vite.

**Reply by tomerzcod7** [reply]
@estevaolucas Have you managed to make this work? :)

**Comment by JimmayVV** [comment]
you have no idea how relevant this to our current needs. Very hopeful this happens.

**Comment by v-anton** [comment]
This is exactly what I wanted to have for our Vite+RR app! Looking forward

**Comment by sergiodxa** [comment]
This could open the door to a SSG mode in the future that just do SSR and fetch a set of URLs to generate static HTML inside the public folder. It could work in hybrid mode by removing the static routes from the manifest.

**Reply by ryanflorence** [reply]
#7641 Will help with that too by changing from a `?_data={routeid}` param to the `{url}.json` extension. Then you can write a simple script to pre-generate all the json files in public at the URLs the remix runtime will request them.

**Reply by sergiodxa** [reply]
Yes, a combination of both HTML and the JSON will make this simpler.

**Reply by ryanflorence** [reply]
Yeah exactly. #7641 is critical here not just for the file extension but the coupling of documents and json makes it all so much more simple (for remix and consumers):

```tsx
import path from "path";
import { createRequestHandler } from "@remix-run/server-runtime";
import * as build form "./build"

let handler = createRequestHandler(build);

let hrefs = ["/", "/about", "/contact"];

for (let href of hrefs) {
  let [html, json] = await Promise.all([
    handler(href).then(r => r.text()),
    handler(href + ".json").then(r => r.json())
  ]);

  if (href === "/") href = "index"

  await fs.mkdir(path.dirname(href), { recursive: true});
  await Promise.all([
    fs.writeFile(html, `public/${href}.html`),
    fs.writeFile(html, `public/${href}.json`)
  ])
}

console.log("Remix SSG ...");
```

Now browsers and the client side remix runtime will request files that will be served straight out of public/.

**Reply by sergiodxa** [reply]
Something nice is that you can still have a mix of static and dynamic ones, build `/articles/123` Remix will still handle `/articles/456`.

**Reply by rowanc1** [reply]
This is awesome and exactly what we are doing in https://mystmd.org for static site generation for scientific papers. We are currently stuck on v1.17.0 because we are generating these assets and then changing some of the base-paths manually so that they can be deployed on GitHub pages (or in other static folders where additional url folders are put in). There was a change in https://github.com/remix-run/remix/pull/6409 that infinitely reloads the page when the path doesn't match the original.

Hopefully this can take that use case into account, allowing us to deploy simple remix sites that can load JSON and deploy to github pages!

**Comment by grabbou** [comment]
Have you considered sticking to `loaders` and `actions` instead of introducing `clientLoader`? 

Remix could take your `loaders` and `actions`, generate a backend and expose these via an endpoint. Then, in the client bundle, we could substitute that with `fetch` call behind the scenes. This would essentially make it unnecessary to define both clientLoader (with only purpose to fetch data from an endpoint) and separate resource route (to return an actual data from the server). 

I was experimenting with that approach before and I have a POC ready. I was initially working on it as a part of my research bringing RSC and Server Actions into non-server targets (such as React Native). Happy to send you some more information or demonstrate it.

**Reply by sergiodxa** [reply]
> Remix could take your loaders and actions, generate a backend and expose these via an endpoint

This already happens, if you have a route file like `routes/admin.users.tsx` the loader and actions will be available under `/admin/users?_data=routes/admin.users`.\

> Then, in the client bundle, we could substitute that with fetch call behind the scenes

This is also what happens right now? Remix replace your loader and actions with a RR one that just fetch your server loaders and actions.

The purpose here is to let you control this.

> This would essentially make it unnecessary to define both clientLoader (with only purpose to fetch data from an endpoint) and separate resource route (to return an actual data from the server)

You don't need to define a separate resource route in Remix, and this won't make you do it too, you still have loader and action for the server and clientLoader and clientAction for the browser.

**Reply by grabbou** [reply]
Got it, thanks for clarification. 

What happens in traditional Remix application when page is being loaded? My understanding was that loaders run on the server and response is streamed to the client - there’s no “fetching” per se. 

In that sense, I get the distinction between loader and client loader and I find it quite smart that you can explicitly await server loader from the browser (alongside other stuff you may want to do). 

I think given other targets and interoperability between them, it makes sense to have clear distinction between client and server.

**Reply by sergiodxa** [reply]
> What happens in traditional Remix application when page is being loaded? My understanding was that loaders run on the server and response is streamed to the client - there’s no “fetching” per se

You (the app developer) don't fetch it, but a fetch is happening, it's just happens that Remix is the one doing it for you, otherwise there would be no way (aside of doing a document request) to get a route loader's data.

You could say a client loader/action has always existed, but it was just doing a fetch to the correct server loader/action, now you will be able to provide your own client loader/action and decide when to fetch the server one, or even don't do it.

**Comment by ryanflorence** [comment]
Michael and I bike-shedded a bit here, thinking maybe just do a boolean since "spa" is a muddy term.

```tsx
ssr: false
```

We'll also want `vite build --ssr` to throw if set to false.

**Reply by sergiodxa** [reply]
Why the flag then? If Vite already has `--ssr` couldn't use default to SPA target unless SSR is enabled?

**Comment by romansndlr** [comment]
OK, so I'll write down my problem and someone smarter please tell me if this proposal solves it.

I have a Remix app that I want to wrap inside a native shell using Capacitor.js and distribute it to the app stores.

Capacitor supports a server url option but if you enable it, all the assets don't get bundled on the device. 

This causes apple to decline the app from the store saying that there isn't a big enough difference from the web version.

If the server url option is off, Capacitor looks for an index.html file in the public directory.

So in a dream world for me, if you set ssr: false in your Remix config, the build ignores entry.server and the root route?! And supports some way for you to have an index.html file that points to your static assets? 

I don't really know if it can work like this but if someone has and idea, I think being able to move between ssr and no ssr with the same codebase would solve a lot of problems.

**Reply by sergiodxa** [reply]
Why ignore root? You need that to render the <html>, <head> and <body> tags, if any it will generate it statically once.

**Reply by romansndlr** [reply]
Yeah, you could generate the index.html based on your root.tsx at build time for sure.

**Reply by yimingjfe** [reply]
@sergiodxa Is there more information about "generate the index.html based on your root.tsx at build time"?  I don't see it mentioned in the discussion, so if you do, please point me to it, thank you very much!

**Reply by hoffnung8493** [reply]
@romansndlr Try React Native's webview instead of Capacitor. I used to use Capacitor a lot, and got into the exact same issue when I used Remix with Capacitor. Not only does RN have more and better maintained libraries, u dont have to go with the unrecommended serverUrl method with Capacitor. Now I have a admin web app, mobile app and the "backend" running in a single remix codebase!

**Reply by tbscode** [reply]
Anyone having success here?
Having a SSR remix app + SPA capacitor mobile app in one code base would be incredibly nice!

( I know that I can use `clientLoader` only and still get SSR but that has other drawbacks as request aren't made on the server anymore and I also cannot `await serverLoader` anymore )

I think having the option to suppress build errors that are raised then using `loader` or `action` with `ssr: False` would be really cool. @hoffnung8493 Capacitor offers so many native integration that just work cross platform so I don't see React Native webview as a satisfying alternative.

**Reply by hoffnung8493** [reply]
@tbscode For me, the number one requirement was to run Remix with normal SSR on mobile webview. This would be a single codebase for both web and mobile. With `ssr: true`, you suddenly need a separate backend. And to keep it simple, u might even have to use SPA mode for the web app, or u have duplicate backend code.

I have been using Capacitor for many years. With Capacitor to have this true single codebase working, I need to enable `server.url`, where I pass in the production url of the Remix app. But after recent update, it seemed like `server.url` cannot be used in production. Unlike before, in iOS, if u enable `server.url`, the url tab at the top is no more hidden. Maybe there is a way to hide it, but I couldnt find it. Also, the bluetooth packages with Capacitor didnt work well. 

So I was kind of forced to try out RN. Bluetooth problem was immediately resolved with RN. I think RN has a bigger community, eco system etc. Since the RN webview can communicate with RN native app, u get access to all their RN packages, which in general I think are better maintained. Also, u can run Remix with SSR, cuz u can just provide the production remix app's link to the RN webview (instead of localhost). 

And on top of that Expo was an amazing dev experience.

**Reply by tbscode** [reply]
@hoffnung8493 no this is exactly what I'm trying to avoid.
I the whole app to be statically bundled when and that should then be served from the capacitor web view.
I would like the app to have an 'offline' mode, this I've managed to achieved in the past using redux + capacitor + webpack.

It would still be fine if request are routed though a deployed remix server if internet is available, but the assets should be bundled within the app. 

And this does seem to be achievable with `ssr: false`, the `index.html` seem to be render-able without server access, but I cannot use the same code-base as for my web-deployment I'd like to use `loader` functions, which now raises an exception when 'switching' from `ssr: true -> ssr: false`.

So to truly have only one code base for my web deployment and the offline compatible mobile app, I'd need a way to  ignore that exception or toggle use of loader functions.

**Reply by matt-winfield** [reply]
@tbscode Did you get Remix working with CapacitorJS? It seems like an ideal solution if there was some way to ignore `loader` functions:

- Full SSR for web
- SPA for mobile apps, calling fetch for the loader data when needed

**Reply by tbscode** [reply]
Hey @matt-winfield,

No, I was not able to figure it out with Remix. But I did find [vike.dev](https://vike.dev), which is a great Vite-based frontend framework.

It supports all sorts of cool features, but notably it allows using SSR and then just toggling a pre-render flag, which allows for all pages to be pre-rendered (which is, by the way, the multi-entry pre-render feature @donaldpipowitch might be asking for below).

So I basically have a vike.dev + Express + Redux setup now that can be pre-rendered and fully used through a Capacitor web-view (and an offline mode could be added). Part of the setup [is published but still a work in progress here](https://github.com/tbscode/django-vike-chat)

It was quite the daunting setup, though, and Vike can be quite complex. I recommend checking it out, though. Also, Remix had some updates, so maybe things have changed there as well, idk.

**Comment by donaldpipowitch** [comment]
If I understood the proposal correctly just a single HTML file will be generated, correct? Are there plans to create multiple entry HTMLs at some point (to support different initial prerendered layouts) which all act as a valid starting point for the same SPA?

**Comment by bbonamin** [comment]
If I understand correctly, this would allow Remix to be used as-is in browser extensions, which require all templates to be bundled! Great!

**Comment by patjakubik** [comment]
Would it be possible to have SSR for the initial page load and switch to a single-page application (SPA) approach for subsequent navigations?

Our site works as follows: Initially, upon visiting our website, a server-side request to our API fetches the necessary data to render the page. As users browse the site, they directly retrieve additional data from the API, bypassing the server for these subsequent requests.

**Reply by gustavopch** [reply]
Isn’t that just what Remix already does?

**Reply by sergiodxa** [reply]
Yes, but I think he wants to show going to the loader again, this will be possible with client loader and clientAction.

**Reply by patjakubik** [reply]
You're right. It seems that this is already possible. Pardon me!

**Comment by phuoccss** [comment]
Will this and other proposals support having a mix of static routes and SSR routes?
- for 'spa' mode, individual routes can opt-in to SSR mode.
- likewise, for 'ssr' mode, individual routes can opt-in to prerendering.

What about mostly static, prerendered routes with some SSR sections? (eg product page with personalized recommendation)

**Comment by yimingjfe** [comment]
Will the root route only render on the browser side? 

This means that when the project changes from spa mode to ssr mode, I'll have to do some modifications to the root and entry.

**Comment by brophdawg11** [comment]
> The HTML entry point for vite is defined at app/entry.html with a script pointing to entry.client.tsx:

Does the `root.tsx` file get ignored in SPA mode?  Or is it expected to only return the contents of `<div id="app">`?  

I was actually sort of envisioning a world in which we render the `Root` component at build time and dump the output into `index.html`.  That has a few benefits:

* User's can still manage things in the `<head>` with React and leverage `<Meta/>`/`<Links/>` from Remix
* `<Scripts>` would handle rendering the proper build assets with content hashes
* Users can control loading behavior via `clientLoader` and `HydrateFallback` in `root.tsx`
* No changes needed to `root.tsx` to move from SPA to SSR mode

It might veer a bit off from standard vite SPA practice though where the user manages a raw HTML file.

**Reply by brophdawg11** [reply]
This sort of "just works"™️  from a _very_ quick test today:

* Use only `clientLoader`'s in your routes
* Stick a `HydrateFallback` in `root.tsx` (should include `<Scripts>` and probably `<Meta>`/`<Links>` too if you're using them in your routes)
* `npm run dev`
* `curl http://localhost:3000/ -o public/index.html`
* stop dev server
* `npx http-server public`
* Load up http://localhost:8081/

**Comment by ngbrown** [comment]
I read this proposal as primarily having the motivation for ease of migration from React Router SPA to Remix SPA, and eventually to Remix SSR.

In the case of servers where a JavaScript backend isn't possible (e.g. GitHub Pages), I think the Remix documentation should have some guidance for new applications to help them make a choice between React Router SPA and Remix SPA. What are the advantages and disadvantages of starting with one over the other?

**Reply by brophdawg11** [reply]
> eventually to Remix SSR

This is not everyone's goal (nor should it be) - you can stop at Remix SPA

> React Router SPA and Remix SPA

There shouldn't be any functional difference.  Remix SPA is just RR SPA with an included Vite plugin for file-based routing.  Only reason I could see to recommend RR at that point is wanting to use something other than Vite

**Reply by cmacdonnacha** [reply]
Are there any other reasons to use Remix SPA over RR + Vite?

**Reply by sergiodxa** [reply]
@cmacdonnacha you get the file convention from Remix so you don't need to create your routes config yourself, also if you later plan to enable SSR it's just one config and you can start using it, you may also get other features like `<Link prefetch>` (only available in Remix and not in RR)

**Reply by brophdawg11** [reply]
With Remix SPA, you also will be hydrating the document by default (but don't have to) and therefore you get access to Remix's `<Links>/<Meta>` components/functionality and don't need to reach for something like `react-helmet`.

**Comment by AsuraKev** [comment]
Missing features like SPA mode + Module federation is why I am looking at Modern JS framework otherwise I ll be using Remix all the way :)

**Comment by jrestall** [comment]
We use Vite's [library mode](https://vitejs.dev/guide/build#library-mode) to distribute our React Router SPA, rather than an index.html. This is so that the app can be embedded within other sites.

Would the following be possible?

```tsx
// vite.config.js
import { resolve } from 'path'
import { defineConfig } from 'vite'

export default defineConfig({
  plugins: [
    remix({
      target: "spa",
    }),
  ],
  build: {
    lib: {
      entry: resolve(__dirname, 'src/app/entry.client.tsx'),
      name: 'MyApp',
      fileName: 'my-app',
    },
  },
})

```

**Reply by jrestall** [reply]
Actually looking at the [SPA Mode PR](https://github.com/remix-run/remix/pull/8338), I think I can get this working if a small change to the PR is made. If the plugin.ts file can check the Content-Type header as below, then we can generate either a index.js or index.html based on what the server entry handler returns.

```ts
// plugin.ts
...
let response = await handler(new Request("http://localhost/"));
invariant(response.status === 200, "Error generating the index.html file");
let entryFileName =
  response.headers.get("Content-Type") === "application/javascript"
    ? "index.js"
    : "index.html";
// Write out the entry file for the SPA
await fse.writeFile(
  path.join(assetsBuildDirectory, entryFileName),
  await response.text()
);
...
```

Then a custom entry.server.tsx can be used to generate the index.js file rather than a index.html.

```tsx
// app/entry.server.tsx
import type { EntryContext } from "@remix-run/node";

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const manifest = remixContext.manifest;
  let js = `
  import ${JSON.stringify(manifest.url)};
  import * as route0 from ${JSON.stringify(manifest.routes["root"].module)};
  window.__remixContext = ${remixContext.serverHandoffString};
  window.__remixRouteModules = { root: route0 };
  
  import(${JSON.stringify(manifest.entry.module)});`;

  return new Response(js, {
    headers: { "Content-Type": "application/javascript" },
    status: responseStatusCode,
  });
}
```

This will create an index.js file like the following.

```js
// build/client/index.js
import "/assets/manifest-2c6b0849.js";
import * as route0 from "/assets/root-ZzzUgvZw.js";
window.__remixContext = {"url":"/","state":{"loaderData":{"root":null,"routes/_index":null},"actionData":null,"errors":null},"future":{"v3_fetcherPersist":false,"v3_relativeSplatPath":false},"isSpaMode":true};
window.__remixRouteModules = { root: route0 };
  
import("/assets/entry.client-Zxc-DKZW.js");
```

It's not as flexible as Vite's library mode which supports multiple entry points and formats, but this small change would cover most use cases and allow Remix apps to be embedded into other sites. 

Also tip for those testing the PR, you can serve the build output with the npm script,
`"serve": "http-server ./build/client --proxy http://localhost:8080?"`

**Reply by brophdawg11** [reply]
Interesting idea - I'll bring it up internally for discussion.  At the moment, the way I'd recommend doing it is with a small post-processing step to extract the `<script>` contents of `index.html` into an `index.js` file - they'll have exactly what you need and avoid needing any custom stuff in `entry.server` since you are leveraging `<Scripts>` directly via the `<RemixServer>` render.

**Reply by jrestall** [reply]
Thanks @brophdawg11, embedded react widgets and using react for part of an existing page is a common way to use react and react-router as documented at https://react.dev/learn/add-react-to-an-existing-project#using-react-for-a-part-of-your-existing-page so I'm hopeful this can be supported.

**Reply by brophdawg11** [reply]
@mjackson and I talked a bit more about this and I think we're going to skip bringing this over as a first class API.  

Remix SPA Mode is really just a higher level abstraction around React Router + Vite that brings along a few goodies:

1. Filesystem routing
2. Full-document hydration + `<Meta>`/`<Links>` support
3. You are now 1 small config change away from opting into SSR on a route-by-route basis

The concept of "library mode" negates (2) and (3) if you don't manage the HTML and have no interest in SSR so it's likely better to stick to the lower level APIs (RR + Vite) for that use case.  You can also build (1) in userland without too much work as a vite plugin (some folks [already have](https://github.com/zkdiff/vite-plugin-react-router-remix-routes)).  I don't think there's anything else you'd be missing out on using RR/Vite directly, right?

This also helps us keep the surface area manageable from a documentation/communication standpoint.

**Reply by brophdawg11** [reply]
That said if you write the code required to do what I suggested in https://github.com/remix-run/remix/discussions/7638#discussioncomment-8007225 please post it here for other that may be interested in this approach!

**Reply by jrestall** [reply]
Thanks @brophdawg11 , really appreciate you considering this and doing the POC. 

We share feature code between our legacy websites that are incrementally modernizing onto RR 6 + Vite and our modern sites that are on Remix, so I'll look into your suggestion as having everyone on Remix using the same conventions and imports would simplify things.

**Comment by joseph-lozano** [comment]
Sorry if this is a silly question but will API routes still work with in SPA mode?

**Reply by sergiodxa** [reply]
RR allows routes without components so resource routes will most likely keep working, just that client side

They can be useful for using them together with useFetcher

**Reply by joseph-lozano** [reply]
I tried out the preview yesterday. It worked in dev mode, but `build` did not work.

**Reply by brophdawg11** [reply]
If you mean "API routes" as in server resource routes, then no - those will not work in SPA mode since SPA mode doesn't have a server.

You can do a client-side resource route with a `clientLoader` though, just as you would in a React Router app:

```
// routes/api.tsx
export function clientLoader() {
  return "API ROUTE DATA";
}
```

**Comment by brophdawg11** [comment]
Released in `2.5.0` - should flip to stable in the next release or two: https://github.com/remix-run/remix/blob/main/CHANGELOG.md#v250

**Comment by AYastrebtsov** [comment]
Hi everyone! Can somebody clarify for me please?

In February I am going to start develop a new project. It HAS to be an SPA.
We've done a few SSR projects in our company with Remix and we are totally loving it!

For SPA we used React Router + TanStack Query.

But now I see that there is SPA Mode and clientLoaders and clientActions in Remix.

We would love to unify our DX, It would be sooo great if we could use Remix and the same mindset for SSR and SPA projects.

So far I understand this:
1) clientLoader and clientActions are stable and can be used.

2) How risky it is, to start this project with Remix SPA and Vite? From this RFC is seems that Remix SPA it pretty stable, or am I wrong?

Development of this project will take 2-3 months I guess, and we would really like to start this project with Remix SPA. More than that, Vite fixes all problems which we had with our UI library of choice - PrimeReact.

We love Remix very much, thanks everyone for such a great framework, from Russia with love, Andrey.

**Reply by aarjithn** [reply]
Related, I would **love** to have next Remix single to be on SPA mode. Please make it happen @ryanflorence and co.

We have many documentation around "standard" remix, but how we can use SPA mode is very 'underexplored'.

**Reply by brophdawg11** [reply]
Remix SPA will very likely "stabilize" when the Remix vite plugin stabilizes, which is close.  We won't commit to any hard dates but it's getting closer every day.

That said, we should be clear what `unstable_` means here too.  It does not mean "full of bugs" - rather we believe the APIs are well tested, production ready, and without known bugs.  The `unstable_` prefix only means that there _could_ be small breaking changes to the API in subsequent minor releases as we receive more and more feedback from early adopters.  These types of breaking changes will always be called out in release notes, so as long as you're reading release notes closely upon upgrading versions and are not opposed to having to make a few small changes here and there to any `unstable_` APIs - then there's no reason not to start your new project in Vite + SPA Mode IMO

**Reply by AYastrebtsov** [reply]
Thanks for clarification! We'll start our project next week. Happy to be early adopter this time, and will be glad to be of any help!

**Reply by mhmdjaw** [reply]
Theyy just released the stable version of remix vite (including SPA mode)!!! Check their [blog](https://remix.run/blog/remix-vite-stable). If you already started your project with the unstable version you can easily update to the stable one now!

**Comment by matt-winfield** [comment]
Is there a technical limitation that causes `loader` in SPA mode to throw an error instead of a warning? As mentioned in a few comments [here](https://github.com/remix-run/remix/discussions/7638#discussioncomment-8111207) and [here](https://github.com/remix-run/remix/discussions/7638#discussioncomment-7502701), it would be very helpful to be able to build the same app in _both_ SSR mode (for web) and SPA mode (for "bundling" with native apps).

I think there should be an option to make these return a warning instead of an error.

**Reply by brophdawg11** [reply]
As [mentioned above](https://github.com/remix-run/remix/discussions/7638#discussioncomment-8114840) SPA mode is intended to be a simple guardrails-enabled way to develop a SPA.  It's not intended to be used in these more advanced use-cases where you wish to use a combination of the two and do some prerendering.

React Router v7 will have pre-rendering/SSG built-in which will allow you to pre-render using server loaders and then use only client loaders from there on out.

FWIW, you can already do most of this today by pre-rendering your pages manually with `curl` or another crawling approach.  There are multiple example sin https://github.com/brophdawg11/remix-ssg.
